crates/wasmtime/src/runtime/vm/mmap_vec.rs

```diff
@@ -74,6 +80,20 @@ impl MmapVec {
         MmapVec::Alloc { base, layout }
     }
 
+    #[cfg(not(has_virtual_memory))]
+    fn new_externally_owned(ptr: *const u8, len: usize, alignment: usize) -> MmapVec {
+        let layout = Layout::from_size_align(len, alignment)
+            .expect("Invalid size or alignment for MMapVec allocation");
+
+        // NonNull doesn't allow descrimination between mut and non-mut pointers.
+        // `as_mut` on SendSyncPtr is unsafe and to ensure soundness we avoid
+        // its use here.
+        let base = SendSyncPtr::new(
+            NonNull::new(ptr as *mut u8).expect("Attempted to create MmapVec from null pointer"),
+        );
```

Comment:
Reusing `NonNull`/`SendSyncPtr` here and casting the const ptr to mut definitely feels hacky so feedback very welcome on a better approach if one would make sense for this.

Reply:
There is no difference between `*mut T` and `*const T` in Rust's semantics, so it isn't a big deal.

Soft preference for using `.cast::<T>()` and `.cast_{const,mut}()` methods for conversions, since they are a little more explicit and future proof, since they cast just one thing at a time, where `as` can change multiple things.